File: cli/cjtCli/pkg/search/engine.go
package search

import (
	"encoding/json"
	"fmt"
	"math"
	"os"
	"path/filepath"
	"sort"
	"strings"
)

// Document 代表一个向量化的接口元数据
type Document struct {
	ID       string    `json:"id"`       // e.g. "GET /v1/orders"
	Metadata string    `json:"metadata"` // 原始文本摘要
	Vector   []float32 `json:"vector"`   // 高维浮点向量
}

// Result 搜索结果
type Result struct {
	ID      string  `json:"id"`
	Summary string  `json:"summary"`
	Score   float64 `json:"score"`
}

type Engine struct {
	Docs []Document `json:"docs"`
}

func NewEngine(docs []Document) *Engine {
	return &Engine{Docs: docs}
}

// Search 执行 Top-K 向量相似度搜索
func (e *Engine) Search(queryVector []float32, queryText string, limit int) []Result {
	if len(queryVector) == 0 {
		return nil
	}

	var results []Result
	queryTextLower := strings.ToLower(queryText)
	
	// 生成 N-Gram (2字, 3字, 4字窗口) 用于中文字符级模糊匹配
	var nGrams []string
	runes := []rune(queryTextLower)
	for windowSize := 2; windowSize <= 4; windowSize++ {
		if len(runes) < windowSize {
			continue
		}
		for i := 0; i < len(runes)-windowSize+1; i++ {
			nGrams = append(nGrams, string(runes[i:i+windowSize]))
		}
	}

	for _, doc := range e.Docs {
		score := CosineSimilarity(queryVector, doc.Vector)
		metadataLower := strings.ToLower(doc.Metadata)
		idLower := strings.ToLower(doc.ID)

		// 补丁逻辑 (N-Gram Hybrid Search): 基于语义碎片命中率加分
		if len(nGrams) > 0 {
			hitCount := 0
			for _, gram := range nGrams {
				if strings.Contains(metadataLower, gram) || strings.Contains(idLower, gram) {
					hitCount++
				}
			}
			// 根据命中比例加分，最高加 0.98
			boost := float32(hitCount) / float32(len(nGrams)) * 0.98
			if boost > 0 {
				score += boost
			}
		}

		if score > 0.3 { // 相似度阈值
			results = append(results, Result{
				ID:      doc.ID,
				Summary: doc.Metadata,
				Score:   float64(score),
			})
		}
	}

	sort.Slice(results, func(i, j int) bool {
		return results[i].Score > results[j].Score
	})

	if len(results) > limit {
		results = results[:limit]
	}
	return results
}

// CosineSimilarity 计算两个向量的余弦相似度
func CosineSimilarity(v1, v2 []float32) float32 {
	if len(v1) != len(v2) || len(v1) == 0 {
		return 0
	}

	var dotProduct, mag1, mag2 float32
	for i := 0; i < len(v1); i++ {
		dotProduct += v1[i] * v2[i]
		mag1 += v1[i] * v1[i]
		mag2 += v2[i] * v2[i]
	}

	if mag1 == 0 || mag2 == 0 {
		return 0
	}

	return dotProduct / (float32(math.Sqrt(float64(mag1))) * float32(math.Sqrt(float64(mag2))))
}

// RebuildIndexFromSpec 从 OpenAPI 数据构建语义搜索索引
func RebuildIndexFromSpec(specData interface{}, indexPath string, embedder func(string) []float32) error {
	spec, ok := specData.(map[string]interface{})
	if !ok {
		return fmt.Errorf("invalid spec data format")
	}

	paths, ok := spec["paths"].(map[string]interface{})
	if !ok {
		return fmt.Errorf("spec does not contain paths")
	}

	var docs []Document
	for path, methods := range paths {
		for method, detail := range methods.(map[string]interface{}) {
			summary := ""
			if d, ok := detail.(map[string]interface{}); ok {
				if s, ok := d["summary"].(string); ok {
					summary = s
				}
			}
			text := strings.TrimSpace(summary + " " + path)
			
			// 调用推理机生成向量
			vector := embedder(text)

			docs = append(docs, Document{
				ID:       fmt.Sprintf("%s %s", strings.ToUpper(method), path),
				Metadata: summary,
				Vector:   vector,
			})
		}
	}

	engine := NewEngine(docs)
	raw, err := json.Marshal(engine)
	if err != nil {
		return err
	}

	os.MkdirAll(filepath.Dir(indexPath), 0755)
	return os.WriteFile(indexPath, raw, 0644)
}

func LoadEngine(indexPath string) (*Engine, error) {
	data, err := os.ReadFile(indexPath)
	if err != nil {
		return nil, err
	}
	var e Engine
	if err := json.Unmarshal(data, &e); err != nil {
		return nil, err
	}
	return &e, nil
}

File: cli/cowen/crates/cowen-ai/src/engine.rs
use serde::{Deserialize, Serialize};
use std::cmp::Ordering;

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Document {
    pub id: String,
    pub metadata: String,
    pub description: String,
    pub vector: Vec<Float32>, // Using float32 for consistency with Go
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SearchResult {
    pub id: String,
    pub summary: String,
    pub description: String,
    pub score: f64,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Engine {
    pub docs: Vec<Document>,
}

impl Engine {
    pub fn new(docs: Vec<Document>) -> Self {
        Self { docs }
    }

    pub fn search(&self, query_vector: &[f32], query_text: &str, limit: usize) -> Vec<SearchResult> {
        if query_vector.is_empty() {
            return Vec::new();
        }

        let query_text_lower = query_text.to_lowercase();
        let query_runes: Vec<char> = query_text_lower.chars().collect();
        let mut n_grams = Vec::new();

        for window_size in 2..=4 {
            if query_runes.len() < window_size {
                continue;
            }
            for i in 0..=(query_runes.len() - window_size) {
                n_grams.push(query_runes[i..(i + window_size)].iter().collect::<String>());
            }
        }

        let mut results = Vec::new();
        for doc in &self.docs {
            let mut score = cosine_similarity(query_vector, &doc.vector);
            
            let metadata_lower = doc.metadata.to_lowercase();
            let desc_lower = doc.description.to_lowercase();
            let id_lower = doc.id.to_lowercase();

            if !n_grams.is_empty() {
                let mut hit_count = 0;
                for gram in &n_grams {
                    if metadata_lower.contains(gram) || desc_lower.contains(gram) || id_lower.contains(gram) {
                        hit_count += 1;
                    }
                }
                
                let boost = (hit_count as f32) / (n_grams.len() as f32) * 0.98;
                if boost > 0.0 {
                    score += boost;
                }
            }

            if score > 0.3 {
                results.push(SearchResult {
                    id: doc.id.clone(),
                    summary: doc.metadata.clone(),
                    description: doc.description.clone(),
                    score: score as f64,
                });
            }
        }

        results.sort_by(|a, b| b.score.partial_cmp(&a.score).unwrap_or(Ordering::Equal));
        if results.len() > limit {
            results.truncate(limit);
        }
        results
    }
}

pub fn cosine_similarity(v1: &[f32], v2: &[f32]) -> f32 {
    if v1.len() != v2.len() || v1.is_empty() {
        return 0.0;
    }

    let mut dot_product = 0.0;
    let mut mag1 = 0.0;
    let mut mag2 = 0.0;

    for i in 0..v1.len() {
        dot_product += v1[i] * v2[i];
        mag1 += v1[i] * v1[i];
        mag2 += v2[i] * v2[i];
    }

    if mag1 == 0.0 || mag2 == 0.0 {
        return 0.0;
    }

    dot_product / (mag1.sqrt() * mag2.sqrt())
}

// Support type alias since f32 is float32 in Go but f32 in Rust.
// But some crates like ndarray or ort might prefer f32.
type Float32 = f32;
